src/content/en/updates/2018/06/audio-worklet-design-pattern.md

```diff
@@ -0,0 +1,414 @@
+regular AudioNode processes 128
+[frames](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API/Basic_concepts_behind_Web_Audio_API#Audio_buffers_frames_samples_and_channels)
+at a time, AudioWorkletProcessor must do the same to become a first-class
+citizen. This is one of the advantages of the AudioWorklet design that ensures
+no additional latency due to internal buffering is introduced within the
+AudioWorkletProcessor, but it can be a problem if a processing function requires
+a buffer size different than 128 frames. The common solution for such case is to
+use a ring buffer, also known as a circular buffer or a FIFO.
+
+Here's a diagram of AudioWorkletProcessor using two ring buffers inside to
+accommodate a WASM function that takes 512 frames in and out. (The number 512
+here is arbitrarily picked.)
+
+<figure>
+  <img
+      class="diagram"
+      src="/web/updates/images/2018/06/aw-design-pattern-4.png">
+  <figcaption>
+    Using RingBuffer inside of AudioWorkletProcessor's `process()` method
+  </figcaption>
+</figure>
+
+The algorithm for diagram would be:
+
+  1. AudioWorkletProcessor pushes 128 frames into the **Input RingBuffer** from
+    its Input.
+  1. Perform the following steps only if the **Input RingBuffer** has greater
+    than or equal to 512 frames.
+     1. Pull 512 frames from the **Input RingBuffer**.
+     1. Process 512 frames with the given WASM function.
+     1. Push 512 frames to the **Output RingBuffer**.
+  1. AudioWorkletProcessor pulls 128 frames from the **Output RingBuffer** to
+    fill its **Output**.
+
+As shown in the diagram, Input frames always get accumulated into Input
+RingBuffer and it handles buffer overflow by overwriting the oldest frame block
+in the buffer. That is a reasonable thing to do for a real-time audio
+application. Similarly, the Output frame block will always get pulled by the
+system. Buffer underflow (not enough data) in Output RingBuffer will result
+silence causing a glitch in the stream.
+
+This pattern is useful when replacing ScriptProcessorNode (SPN) with
+AudioWorkletNode. Since SPN allows the developer to pick a buffer size between
+256 and 4096 frames, so the drop-in substitution of SPN with AudioWorkletNode
+can be difficult and using a ring buffer provides a nice workaround. A audio
+recorder would be a great example that can be built on top of this design.
+
+However, it is important to understand that this design only reconciles the
+buffer size mismatch and it does not give more time to run the given script
+code. If the code cannot finish the task within the timing budget of render
+quantum (~3ms at 44.1Khz), it will affect the onset timing of subsequent
+callback function and eventually cause glitches.
+
+Mixing this design with WebAssembly can be complicated because of memory
+management around WASM heap. At the time of writing, the data going in and out
+of WASM heap must be cloned but we can utilize HeapAudioBuffer class to make
```

Comment:
"cloned but" -> "cloned, but"?

"utilize HeadAudioBuffer" -> "utilize the HeapAudioBuffer"